### ISSUE TITLE
Create graphs in R for reproducible research: suggestions and work already done?

### ISSUE THREAD

**Issue by samthiriot** [openmole/openmole#476]
Dear beloved OpenMole community,

as part of the general principles of reproducible research, for which OpenMole is so much valuable, the generation of graphs and other descriptive statistics would better be part of the automated chain of experiments. 

It might also be a way to double check easily what we do. 
For instance when using NSGA2, I would appreciate a plot of the Pareto front. 
When using sensitivity analysis, I would like to plot the corresponding results, etc.

I start developing a R script based on `ggplot2` and ``plotly` to draw these types of plots. 
Suddenly I discover plotly scala in openmole repositories, so I wonder if I'm not reinventing the wheel there. 

Do you already have a way to create plots of simulation experiments? 
Do you have recommendations on how to better do it? 
Anyway, I would of course share my graphs back to the community using the OpenMole market. 

Tks!

Sam

**Comment by JusteRaimbault** [comment]
There was a stream of discussion during the last coding camp in June to kind of standardize what is done with outputs; @romainreuillon has developed some metadata exports and @chapinux @ElCep a R package to do standard plots, and it was linked to the developments of @mathieuleclaire for the plots in the GUI - so my guess is that your contribution to the discussion and implementation would be welcomed; I'm not sure though we have yet a common document to gather what is done on this subject and coordinate

**Comment by samthiriot** [comment]
Thanks Juste, 

interesting! 
Is there already code somewhere so I can understand the general direction? 
Anyway I'm available to contribute this effort. 

Sam

**Comment by chapinux** [comment]
Hallo @samthiriot 

Indeed, we tried with @ElCep  to craft a basic shiny app that takes an "OpenMOLE export tgz file" whatever that means and make the possible useful plots out of it. It was harder and far less fun and quick than expected, even in a cool context like the OM CodingCamp 

For now we just try to mimic the display of OM GUI,  but the long term goal(s) would be  :

- the display should be automagically customized to the type of method, just as you said (e.g. Pareto front for multi criteria) 
- it should run outside OpenMOLE for people who don't want or can't use OM  (Yes , there are such people. Lucky them ) 
- it should be fluid, eye candy (hence the plotly-costly JS +  HTML final object), and taylored for interactive exploration of results  
- It should be "usable/shareable" ,hence the R shiny app that could be packaged as a standalone app later, and the tgz results export format 


The choice of R comes from the fact that we know a little bit of it and the "On peut tout faire avec R"   mantra
Nevertheless, that was a lot of text for so little code ,  here is the repo : [https://github.com/chapinux/RopenMOLE](https://github.com/chapinux/RopenMOLE)

We planned to advance it "when we have time" i.e. never ... until the next Coding Camp !

**Comment by samthiriot** [comment]
Hello @chapinux 

thanks for the story! I'll check it out in detail.

I guess we would have to go step by step. The long term goal of a generic plotting solution is of course the best, but as any generic solution, it sounds like "target the moon"; it often stops when we realized how difficult it is. At least on my side, i'll need short term graph solutions, so I'll go for an incremental approach.

I start for instance with a very simple, really stupid first step: I would like to add in the market test functions for user to test by themselves NSGA2. This requires plotting the last Pareto front so the user can compare the front with the reference graph. 
They can make it already manually by navigating to the last CSV result and configuring the plot in the web GUI; but other cases obviously require more detailed plots. 
 
In a first attempt I produce a PNG from R, the kind of simplest, easiest case. 
(it's waiting for a merge there: https://github.com/samthiriot/openmole-market/tree/test_functions1/nsga2-test-functions )

In general visualizing a PNG from the GUI would be nice, as it can convey a lot of information (ggpairs plot, etc.). 
Would it be possible to display the PNG in the OpenMOLE GUI, or should we just let the user download it? 

I've other more complicated versions producing videos of evolution in mp4, and interactive plotly. Unfortunalely all of that requires a download from the user. I'll post these as examples in the openMole market as a basis for discussion. 

Let's graph :-) 

Sam

**Comment by mathieuleclaire** [comment]
Dear @samthiriot,
Of course, we can add a png viewer in OpenMOLE, but I am not sure to understand the purpose. Why vizualize a png in OpenMOLE if you need to use an external applicaton to produce it ? Note that we plan to produce relevant vizualization depending on the method, giving information on the convergence and the outputs. Your plot attempts are welcomed of course !

Mathieu

**Comment by samthiriot** [comment]
Morning @mathieuleclaire 

Oh, we can easily produce png images using the `RTask`. For instance if you have a set of inputs / outputs, you can plot `ggpairs` easily to get some grasp on the one-to-one dependancies of variables with each other (inputs and outputs). So you can definitely produce PNG from OpenMole. You can check you the [recent demo I shared on the market](https://github.com/openmole/openmole-market/tree/12-dev/nsga2-test-functions) in which a RTask takes the most recent generation file and plots the objectives space as a PNG.  

Actually the simplest way to visualize the evolution of an exploration with a genetic algo, as far as I found yet, is to graph every CSV file representing a generation as a PNG. The user can thus browse and visualize every generation. Based on that you can very easily produce a mp4 using `ffmpeg`, still from within a `RTask` as you can install various soft in the R container using the install parameter. MP4 has for benefits a very small size (thanks to a good compression) so people can both visualize and share their results in a presentation or so. Still the user has to download the mp4 to view it. 

Yet the best would be the interactive plotly as mentioned earlier by @chapinux in the discussion. Using the RTask there is no difficulty producing it, with the obvious benefits of elegance, interactivity. The resulting graphs are also more informative, as a the tooltip of every point in the parameter space contains information about the corresponding parameters. I produced it, but the user has to download the resulting html file, and the associate directory, to open it in its own computer and browser. The graphs can also be slow to open (up to 20Mb in the html page easily, as there are many points for an exploration and the storage format used by plotly is not that efficient). A plotly graph is also not easy to embbed into a presentation or publication. 

So in short I've the feeling the different formats have their own pros and cons. As far as I understand formats such as PNG or MP4 can be of use. I did not mentioned eps and pdf, which are required for the compilation of latex articles. Of course it would be very important and nice to have an interactive visualization integrated inside OpenMole, but that seems complementary. 

What do you think? 

Sam

**Comment by mathieuleclaire** [comment]
Oh, you do it with RTask. Could you give us examples of png and mp4 so that I do no need to run them ;) ? We could get inspiration to build automatic vizualizations.

**Comment by samthiriot** [comment]
Hello @mathieuleclaire ,

here are a few examples of what can easily be done with R. It's just a very first attempt, I don't claim it at all as any desirable result :-) 

Example of the graph of a 2D Pareto front during the evolution
![iteration0013](https://user-images.githubusercontent.com/2084275/91322601-38bf0000-e7c0-11ea-9e5c-1953bc399df7.png)

A "pairs" graph showing the relationships between variables: 
![pairs](https://user-images.githubusercontent.com/2084275/91322554-2e046b00-e7c0-11ea-80ff-2dd27ecf1cde.png)

Example of an mp4 video to understand or show the evolution (had to zip because of github restrictions)
[iterations_video.mp4.zip](https://github.com/openmole/openmole/files/5130736/iterations_video.mp4.zip)

Example of a basic interactive plotly graph
[evolutionPlotly.zip](https://github.com/openmole/openmole/files/5130800/evolutionPlotly.zip)

Again, it's only the very first examples of what we can quickly do with the basic R features. 
I can keep you posted when we get something better ;-)

Sam

**Comment by samthiriot** [comment]
Hello,

For information I shared the code for these basic examples in the market: 
https://github.com/openmole/openmole-market/pull/17

Sam

**Comment by samthiriot** [comment]
Morning @chapinux ,

I read you message again. I bet we are on the same goals for interactive diagrams. I quickly tested R + plolty + shiny and was amazed by the result. 

Here are a few additional thoughts on the topic

- we have to deal with many simulation results, meaning many points; in such a case a pure plotly solution becomes very slow, as all the points are transmitted to the browser for rendering. In a client/server approach like shiny, only the part of the data to be plot is transmitted. If required, some plots can even be rendered on the server side. 
- displaying the Pareto front automatically sounds nice. Yet in practice, as soon as the Pareto front has a higher dimensionality than 2, it becomes more tricky. When you have up to 4 goals (more is anyway not that relevant even for the evolutionary algo), you need a graphing solution which all allows you to change the variables displayed on the graph. So "automatic" sounds like "parameterable" with suitable default settings when there are 2 goals.
- a key question for optimisation is, when we see the space of goals, to create a link with the space of parameters. Ideally we should be able to view both the plot of the results and the plot of the inputs; then when we select a part of the space of results, we might select the corresponding points in the space of inputs
- interactive diagram are nice, but it is important to also have a "save" button which enables the transmission of the knowledge displayed in the graph 

Well, that really are *random* thoughts ^^

Sam

**Comment by samthiriot** [comment]
Morning @all,

Did you ever used CobWeb plots? I don't mean it in the sense of the first google results nor wikipedia pages; another understanding of a CobWeb plot in uncertainty analysis.

This example comes from the OpenTurns developed by my colleagues for uncertainty analysis. 
![example from OpenTurns](http://openturns.github.io/openturns/1.12/_images/VisualTest_DrawCobWeb.png)

On such a graph you can display as many variables as required (vertically), and give a feeling, a bit like in a matrix scatter plot, of the dependancy between variables. 
We might also display discrete variables.
You can condition it, as displayed on this image, meaning the user selects a domain of interest for one or more variables, then only the trajectories (that is, the combination of parameters) matching these conditions are displayed. It answers the question "what combinations of parameters bring to this result?" or the opposite "what results are brought by this combination of parameters?"

Unfortunatly I don't find any library able to display it quickly. I'll dig in the OpenTurns code which is opensource. 

What do you think of this type of diagram?

Sam

**Comment by chapinux** [comment]
Wesh wesh @samthiriot ,

quick answer, as we are on the run for eX Modelo beginning next week, 

-no cobweb experience for me, so no opinion for now :-)

- graphical link  between a user-input selection in output space , and  the selection origin in input space is possible almost freely with plotly and the 'highlight' black magic. see for example : [https://plotcon17.cpsievert.me/workshop/day2/#36](https://plotcon17.cpsievert.me/workshop/day2/#36)

-the problem of size for the combination of many input/output in a html file is real pain in the *ss. Just as you said it can be overcom by streaming  data with more savy coding than usal and I don't really master this kind of things  :cry:, 

- also it rises the question : will it still be packageable as a standalone app for people who want to explore results on their own , without R or OM at hand. (This constraint was at the origin of the RopenMole vizu app, targetted to companion modeling, the IRD/ComMod way.   )

As you may have guessed we are hot like french fries mansions to test these propositions, you definitely need to come to the next OM black sabbat, I mean Coding Camp.

Cheers!

**Comment by samthiriot** [comment]
Thanks @chapinux,

hope the ex Modelo goes well!

I progressed quiet a bit on the visualization topic.
Plotly is definitely an interesting tool and approach.

Here are elements:
* a live demo of the current state of the visu app: https://samthiriot.shinyapps.io/plot-evolution/
* the repository with the sourcecode in LGPL: https://github.com/samthiriot/plot-evolutionary-algo
* the usage from OpenMOLE: https://github.com/samthiriot/plot-evolutionary-algo/tree/master/openmole

There are two distincts parts.
 
One one side, a little R/Plotly/Shiny app which displays the content of a CSV file which contains a column with the iteration id, columns with parameters and columns with goals. Then the app proposes a graphical user interface to explore the data, as you can see in the live demo. There are many thinks to improve, suggestions are welcomed in the bugtracker, collaborative development would be even nicer during the next coding camp :-) 

One the other side, there is the question of the bridge between OpenMOLE and the visu app. 
* A possible way to do the interface is to create an OpenMole task which merges all the CSV files produced by an optimization, then  let the user download the CSV file and display it into the app. I did that in R, but it would be better done in Scala; we might even imagine a dedicated Hook able to merge the CSV files.  
* A more interesting way (to me) is to create an OpenMole task which exports the CSV file and the visu app into a shiny app. Then this shiny app can be downloaded into a zip, then ran quickly from command line, transmitted to other people, etc. I did that with a RTask as shown in the repo, but this might be better done as an OpenMOLE hook. 

What I feel sad is to totally disconnect from OpenMOLE. This should be more easy for the user. 
A possibility might be to have the OpenMOLE web GUI detecting when a zip is a Shiny app, so OpenMOLE starts it into R (why not in a dedicated container to ensure the presence of R and related packages), and OpenMole can open the corresponding web interface with the port of interest. This workflow would be more fluid to the user. 

Well,  I guess the best would be to discuss all of that of loud voice to exchange. 

A bientôt!

Sam

**Comment by JusteRaimbault** [comment]
This is circling back to the old idea of having notebooks (whatever language, Jupyter e.g.) either within OpenMOLE or launched besides with a communication between the two; not sure this is the best path compared to developing the GUI plots which are more and more advanced and ergonomic (to have "une plateforme intégrée" as they said back in the days). The similar platform DAFNI is using Jupyter notebooks plugged as a task in their workflow, so this may be a reasonable option (see https://docs.secure.dafni.rl.ac.uk/ ).

**Comment by samthiriot** [comment]
Thanks @JusteRaimbault ,
An additional aspect on this debate: the users I'm working with already ask for adaptations of GUI matching their specific case (for instance if a simulation produces a geographic map, view the map; if the simulation produces curves, display them, etc.). So I guess we should clearly distinguish what the basic visualization features of OpenMole are - and as evolutionary algorithms are part of the core features of OpenMOLE, to make the integrate visualisation better - and what users can easily develop as specific GUI. I think both the facets are important.

**Comment by mathieuleclaire** [comment]
Hi @samthiriot,
adding image viewer in OpenMOLE is easily doable. 
Plotting user home-made plots seem to be much more harder because it breaks the generic aspect of the plot section. What would do you have in mind ? Building specific plot plugins to visualize specific data representations ?

**Comment by samthiriot** [comment]
Hi @mathieuleclaire 

Incidentally, a very small change proposal for the existing plotting features in OpenMOLE: when you display a X,Y scatter plot, select by default the two last colums of the CSV file as X and Y. In the case of an optimization with two objectives (or 1 objective with stochasticity) this directly gives the user the Pareto front without additional click, and without needing any metadata. The current behavior is to select the same column for X and Y, which is never informative. It's a really small change with a huge benefit for the user :-) 

I would also be nice to facilitate the viewing of novel files in the working directory; when a hook writes a file in the working environment, it might trigger a refresh of the corresponding directory. Else the user has to remember he has to click on "refresh". 

Back to your question. I guess home made plots would be done using a generic framework such as Shiny as discussed before with @chapinux . Advanced visualization requires specific languages, libraries, all of that changing quickly; I think it is better to rely on an existing solution rather than trying to reinvent a configurable visualization. So to me the message to a user would be: _"here are the built in visualization features of OpenMOLE. They allow you to understand convergence of your exploration, visualize your Pareto front, etc. If you want to built specific visualizations fitting your specific model, you can just use Shiny_ [? and/or other frameworks]_. Just go to the web and learn how to do it. Now OpenMOLE will make your life easier because everytime an exploration is ran, you just do_ [??? to be defined] _to watch the results using your home made visualization"_ 

I'm not yet able to define a good way to interface a visualisation developed by the user, and OpenMOLE running the experiments. We will try with what I wrote, then I'll be able to give you a REX. What I forsee so far is:
* level 0: have OpenMole exporting one huge CSV file rather than a collection, thanks to a different Hook. Then the user can easily download this file on his computer and use whatever visualization he wants. 
* level 1: write an OpenMOLE task which packages the data from an exploration (the CSV files) and a plotly visualization written by the user in R/plotly. The resulting plotly app (a zip with data and the code for visualization) can be dumped to a zip file in the working directory using a hook. The user can download it and run it in its own environment. This matches the need of producing independant visualization mentioned by @chapinux 
* level 2: once the plotly app was built (level 1), make OpenMole facilitate its execution. It might be: 
    * (a) when a zip is available in the workdirectory, propose in the GUI to run it (double click on it for instance) and open the corresponding page; the user still has to do a step manually, but it is a matter of one click.
    * (b) propose OpenMOLE hooks or tasks which take a Shiny app (the zip file) and run it in a container; or upload it to a cloud running shiny apps (such as shinyapps.io); this way the user does what he wants. With such a workflow, the user can execute automatically his home made visualization at every run, yet this does not requires a change in the architecture of OpenMOLE.

Another question: should visualization be only done after the entire exploration? Wouldn't it be better to watch the evolution during the exploration so we know if convergence what reached? 

 Pleasure to discuss more,

**Comment by samthiriot** [comment]
Hello,
on the side of generic visu, I adapted the live demo of the visu: https://samthiriot.shinyapps.io/plot-evolution/ 
As this is just a combination of standard Shiny/Plotly widgets,I guess this type of visu can be integrated inside OpenMOLE at a reasonable cost, in the same way the current visualizations are made. 
Sam
